tokio/src/runtime/mod.rs

```diff
@@ -435,13 +436,49 @@ impl Runtime {
     /// ```
     ///
     /// [handle]: fn@Handle::block_on
-    pub fn block_on<F: Future>(&mut self, future: F) -> F::Output {
-        let kind = &mut self.kind;
-
-        self.handle.enter(|| match kind {
-            Kind::Shell(exec) => exec.block_on(future),
+    pub fn block_on<F: Future>(&self, future: F) -> F::Output {
+        self.handle.enter(|| match &self.kind {
+            Kind::Shell(exec) => {
+                // TODO(lucio): clean this up and move this impl into
+                // `shell.rs`, this is hacky and bad but will work for
+                // now.
+                let exec_temp = {
+                    let mut lock = exec.lock().unwrap();
+                    lock.take()
+                };
+
+                if let Some(mut exec_temp) = exec_temp {
+                    let res = exec_temp.block_on(future);
+                    exec.lock().unwrap().replace(exec_temp);
+                    res
+                } else {
+                    self.handle.enter(|| {
```

Comment:
why's the second call to `enter` necessary here?

Reply:
This is the `enter` for the handle context the next is for the runtime enter to avoid a double enter iirc. This is just a port of what the code was doing in `Handle::block_on` iirc.

See https://github.com/tokio-rs/tokio/blob/d9d909cb4c6d326423ee02fbcf6bbfe5553d2c0a/tokio/src/runtime/handle.rs#L274

Reply:
I'm not following why we need two enters.

Reply:
yeah, im blind, I thought you were referring to the `runtime::enter`, will fix this 👍